slice/Ice/Exceptions.ice

```diff
@@ -63,4 +63,10 @@ module Ice
         /// The name of the corresponding parameter.
         string paramName;
     }
+
+    /// A marshal exception is thrown when a request or response can't be encoded. This can occur if the
```

Comment:
When a request cannot be written, Ice should throw a local exception (that is, a System.Exception that is not a ZeroC.Ice.RemoteException). Not being able to write a request is purely local. When your invocation fails with a remote exception, the expectation is the request was sent. (Therefore, when LocatorInfo fails to resolve an endpoint, it should not throw or let through a remote exception, but that's a different topic).

So here the remote MarshalException should be used only when Ice fails to marshal a response or exception? For the future Data frames from server to client, I assume a marshaling failure just resets the stream and does not send an exception.

I am also wondering if MarshalException is the best exception name here.

We already have UnhandledException for unexpected errors such as:
 - failing to unmarshal request parameters
 - the user code throws an exception that is not a RemoteException
 - the user code throws an remote exception with ConvertToUnhandled set to true

UnhandledException represents a bug in the application, so if we consider that when the marshaling of a response or exception exceeds the peer's IncomingFrameMaxSize is not a bug but "expected", it's desirable to use a different remote exception. Maybe `LimitExceededException` - an exception thrown by the remote peer when some limit is exceeded? Or `ImplementationException` - an exception thrown by the remote peer's implementation that is not unhandled, so may not be a bug?

Reply:
Before you suggested to use MarshalException, the implementation originally raised ArgumentOutOfRange exception if the request couldn't be written when too large. What local exception do you suggest to raise here instead? To me ArgumentOutOfRangeException or ArgumentException with an appropriate message still works since the exception is the result of arguments being too large to encode.

I don't have a strong preference for the remote exception... I don't think UnhandledException is necessarily bad either if we want to keep it simple. To me the line is quite thin on whether or not we should consider this as an application bug or not. So what do you prefer?